### PR TITLE
Add analyzer for comparing ISymbol in a handful of collection operations

### PR DESCRIPTION
**Pull request by ryzngard**
Methods that are flagged are listed below. This was chose in favor of searching for applicable constructors or methods that take an IEqualityComparer to reduce performance hit. 

### `System.Linq.Enumerable<ISymbol>`
- Contains
- Distinct
- GroupBy
- GroupJoin
- Intersect
- Join
- SequenceEqual
- ToDictionary
- ToHashSet
- ToLookup
- Union

### `Dictionary<ISymbol, T>`
- Constructor

### `HashSet<ISymbol>`
- Constructor

cc @chsienki  @jcouv

### CODE REVIEW

**Comment by ryzngard** [comment]
ping @mavasani @sharwell

File: src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx [hunk 1]
```diff
@@ -374,4 +374,13 @@
   <data name="EnableConcurrentExecutionFix" xml:space="preserve">
     <value>Enable concurrent execution</value>
   </data>
+  <data name="UseComparerInSymbolCollectionsTitle" xml:space="preserve">
+    <value>Use a comparer in symbol collections</value>
+  </data>
+  <data name="UseComparerInSymbolCollectionsMessage" xml:space="preserve">
+    <value>When creating collections or using methods that accept IEquailityComparer one should be provided</value>
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx]: https://www.bing.com/images/search?q=quail+comparison

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/UseSymbolComparerInCollectionsAnalyzer.cs [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            description: s_localizableDescription,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/UseSymbolComparerInCollectionsAnalyzer.cs]: This is a correctness issue, and it's unlikely that we'll face code generators that both do the wrong thing _and_ cannot be updated to behave correctly.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/UseSymbolComparerInCollectionsAnalyzer.cs [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var compilation = context.Compilation;
+                var symbolType = compilation.GetTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisISymbol);
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/UseSymbolComparerInCollectionsAnalyzer.cs]: I filed #2953 to head this off.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/UseSymbolComparerInCollectionsAnalyzer.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ryzngard) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/UseSymbolComparerInCollectionsAnalyzer.cs]: Analyzing generated code was unintentional, just a copy-paste remnant

File: src/Microsoft.CodeAnalysis.Analyzers/UnitTests/MetaAnalyzers/UseSymbolComparerInCollectionsTests.cs [hunk 4]
```diff
@@ -0,0 +1,618 @@
...
+    Microsoft.CodeAnalysis.Analyzers.MetaAnalyzers.Fixers.UseSymbolComparerInCollectionsFix>;
+using VerifyVB = Test.Utilities.VisualBasicSecurityCodeFixVerifier<
+    Microsoft.CodeAnalysis.Analyzers.MetaAnalyzers.UseSymbolComparerInCollectionsAnalyzer,
+    Microsoft.CodeAnalysis.Analyzers.MetaAnalyzers.Fixers.UseSymbolComparerInCollectionsFix>;
+
+namespace Microsoft.CodeAnalysis.Analyzers.UnitTests.MetaAnalyzers
+{
+    public class UseSymbolComparerInCollectionsTests
+    {
+        private const string SymbolEqualityComparerStubCSharp =
```
Comment (sharwell) [inline src/Microsoft.CodeAnalysis.Analyzers/UnitTests/MetaAnalyzers/UseSymbolComparerInCollectionsTests.cs:18]: 📝 This is no longer needed. The `ReferenceAssemblies` can be specified to test both Roslyn 2.x (no `SymbolEqualityComparer`) and Roslyn 3.x.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/UseSymbolComparerInCollectionsAnalyzer.cs [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+            {
+                var compilation = context.Compilation;
+                var symbolType = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisISymbol);
+                if (symbolType is null)
+                {
+                    return;
+                }
+
+                var genericComparerType = compilation.GetOrCreateTypeByMetadataName(typeof(IEqualityComparer<>).FullName);
+                var comparerType = genericComparerType?.Construct(symbolType);
```
Comment (sharwell) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/UseSymbolComparerInCollectionsAnalyzer.cs:49]: ❔ What about a set of `ITypeSymbol` or `INamedTypeSymbol`?

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/UseSymbolComparerInCollectionsAnalyzer.cs [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var compilation = context.Compilation;
+                var symbolType = compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisISymbol);
```
Comment (sharwell) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/UseSymbolComparerInCollectionsAnalyzer.cs:42]: 📝 This needs to return early if `SymbolEqualityComparer` is not defined.

**Comment by sharwell** [comment]
Marking as a draft until if/when the merge conflicts can be addressed.

**Pull request closed without merging**